### PR TITLE
fix(shared-decks): can't download decks on Android 15

### LINKED ISSUE #17573: [BUG]: Android 15: Can't download shared decks [This isn't a valid Anki package file / 'deckname.bin']
### Checked for duplicates?

- [x] This issue is not a duplicate

### Does it also happen in the desktop version?

- [ ] This bug does not occur in the latest version of Anki Desktop

### What are the steps to reproduce this bug?

1. Click on "Get shared decks"
2. Download any deck and try to import

https://github.com/user-attachments/assets/7950abce-c333-4cfd-9b48-52f542b72d2d

### Expected behaviour

Deck gets downloaded and imported successfully

### Debug info

```text
AnkiDroid Version = 2.21alpha1-debug (c7b499e62159391f549f73be9a24b2f8c41b4a3f)

Backend Version = 0.1.48-anki24.11 (24.11 c47638ca36f99dd4f3b81ae82d964aec66e392e0)

Android Version = 15 (SDK 35)

ProductFlavor = amazon

Manufacturer = Google

Model = Pixel 6a

Hardware = bluejay

Webview User Agent = Mozilla/5.0 (Linux; Android 15; Pixel 6a Build/AP4A.241205.013; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/131.0.6778.104 Mobile Safari/537.36

ACRA UUID = a432af50-f92f-44f9-979d-b10acb7ea920

FSRS = 1.4.3 (Enabled: false)

Crash Reports Enabled = false
```

### (Optional) Anything else you want to share?

```
10:04:11.845  D  Registering broadcast receiver for download completion
10:04:11.878  D  Download ID -> 982
10:04:11.879  D  File name -> Goethe_Institute_A1_Wordlist.bin
....
10:04:14.889  D  Stopping download progress checker
10:04:14.890  I  File does not have 'apkg' or 'colpkg' extension, abort the deck opening task
10:04:14.890  I  File is not a valid deck, hence return from the download screen
...
10:04:23.998  D  Registering broadcast receiver for download completion
10:04:24.024  D  Download ID -> 983
10:04:24.024  D  File name -> German_Federal_States__Deutsche_Bundeslnder.bin
10:04:24.025  D  Starting download progress checker
...
10:04:27.101  D  Stopping download progress checker
10:04:27.101  I  File does not have 'apkg' or 'colpkg' extension, abort the deck opening task
10:04:27.101  I  File is not a valid deck, hence return from the download screen
10:04:27.106  I  Back pressed when download is in progress, show cancellation confirmation dialog
```

When I downloaded the file from web using a browser, the downloaded filename had extension `.apkg` but when the file was downloaded from app, it had extension `.bin`

### Research

- [x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [ ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by david-allison**
[API 35] `URLUtil.guessFileName`: regression guessing file extension [selects '.bin' when better alternatives exist] - https://issuetracker.google.com/issues/382864232

AnkiDroid fails to open `Goethe_Institute_A1_Wordlist.bin`

## Fixes
* Fixes #17573

## Approach
* repurpose `getFileNameAndExtension` to return a class
* Use this to handle updating the extension of the file to `apkg`

## How Has This Been Tested?
* Unit tests
* Instrumented tests:
  * API 35 emulator
  * API 34 emulator 

## Learning (optional, can help others)
https://issuetracker.google.com/issues/382864232

Android 15 is out

@thedroiddiv is awesome (I knew this already, so I guess this wasn't learning and I just wanted to say thank you!)
 
* `File.createTempFile("ok", ".computer")` fails: the length of the first string must be `>= 3`
  * ⚠️ this is now a URL, maybe we want another test value?

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by lukstbit — APPROVED** [review]
LGTM, thanks!

File: AnkiDroid/src/main/java/com/ichi2/anki/MediaRegistration.kt [hunk 1]
```diff
@@ -27,13 +27,14 @@ import com.ichi2.compat.CompatHelper
...
 import com.ichi2.utils.ContentResolverUtil.getFileName
-import com.ichi2.utils.FileUtil.getFileNameAndExtension
+import com.ichi2.utils.FileNameAndExtension
 import timber.log.Timber
 import java.io.File
 import java.io.FileNotFoundException
 import java.io.FileOutputStream
 import java.io.IOException
 import java.io.InputStream
+import java.lang.IllegalStateException
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/MediaRegistration.kt:37]: I don't actually care, but isn't java.lang.* imported by default? Not used to seeing these as concrete imports. This is more of a language question than anything

File: AnkiDroid/src/main/java/com/ichi2/utils/FileUtil.kt [hunk 2]
```diff
@@ -141,13 +141,29 @@ object FileUtil {
  * @param fileName name of the file, before the '.'
  * @param extensionWithDot extension of the file, with a '.'
  */
-data class FileNameAndExtension(val fileName: String, val extensionWithDot: String) {
+@ConsistentCopyVisibility
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/utils/FileUtil.kt:144]: TIL...

**Review by mikehardy — APPROVED** [review]
explains at least in part why there are so many download/import deck errors! That was mysterious.

File: AnkiDroid/src/main/java/com/ichi2/anki/MediaRegistration.kt (hunk 1 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/MediaRegistration.kt:37]: Interesting. It shouldn't be, but not worth fixing (I always ignore imports except during conflicts)

**Comment by david-allison** [comment]
@mikehardy can we get this out as 2.20.1 at some point?

**Comment by david-allison** [comment]
Google replied saying we can set the `mimeType` to `null` for now

I prefer our solution of forcing the extension: we know we're going to get `apkg` here.

**Pull request merged**